src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -4,21 +4,105 @@
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        _applicationStopping = hostApplicationLifetime?.ApplicationStopping ?? throw new ArgumentNullException(nameof(hostApplicationLifetime));
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach the specified state. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches the specified target state. If the resource
+    /// is already in the target state, the method will return immediately.<br/>
+    /// If the resource doesn't reach one of the target states before <paramref name="cancellationToken"/> is signalled, this method
+    /// will throw <see cref="OperationCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resouce.</param>
+    /// <param name="targetState"></param>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> </param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters",
+                                                     Justification = "targetState(s) parameters are mutually exclusive.")]
+    public Task WaitForResourceAsync(string resourceName, string? targetState = null, CancellationToken cancellationToken = default)
+    {
+        string[] targetStates = !string.IsNullOrEmpty(targetState) ? [targetState] : [KnownResourceStates.Running];
+        return WaitForResourceAsync(resourceName, targetStates, cancellationToken);
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach one of the specified states. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches one of the specified target states. If the resource
+    /// is already in the target state, the method will return immediately.<br/>
+    /// If the resource doesn't reach one of the target states before <paramref name="cancellationToken"/> is signalled, this method
+    /// will throw <see cref="OperationCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resource.</param>
+    /// <param name="targetStates">The set of states to wait for the resource to transition to one of. See <see cref="KnownResourceStates"/> for common states.</param>
+    /// <param name="cancellationToken">A cancellation token that cancels the wait operation when signalled.</param>
+    /// <returns>A <see cref="Task{String}"/> representing the wait operation and which of the target states the resource reached.</returns>
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0026:Do not add multiple public overloads with optional parameters",
```

Comment:
Is there a way we can do this without violating these rules? What are the long term implications?

Reply:
I was told to just suppress these but we can choose to expand to a bazillion overloads instead 😁

Reply:
OK seems like you've chosen the lesser of two evils.